File: 47-permutations-ii/permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def backtracking(i):
            if i==len(nums):
                res.append(nums[:])
                return
            for j in range(i,len(nums)):
                nums[i],nums[j]=nums[j],nums[i]
                backtracking(i+1)
                nums[i],nums[j]=nums[j],nums[i]
        #----------------------------------------------------
        res=[]
        backtracking(0)   # i=0
        ans=[]
        for li in res:
            if li not in ans:
                ans.append(li)
        return ans

File: 47-permutations-ii/permutations-ii.java
class Solution {
    public void swap(int[] nums,int i,int j){
        int temp=nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
    public void backtracking(List<List<Integer>> res,int[] nums,int i){
        if(i==nums.length){
            List<Integer> ans=new ArrayList<>();
            for(int ele: nums){
                ans.add(ele);
            }
            res.add(ans);
            return;
        }

        HashSet<Integer> set=new HashSet<>();     
        for(int j=i;j<nums.length;j++){ // i=0 j=0,1,2 // i=1  j=1,2 //i=2 j=2
            if(set.contains(nums[j])==true){
                continue;
            }
            set.add(nums[j]);
            swap(nums,i,j);  //make the choice
            backtracking(res,nums,i+1); //backtrack
            swap(nums,i,j);  // undo the move
        }
    }

    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res=new ArrayList<>();
        backtracking(res,nums,0);
        return res;
    }
}